elevate-cpanel

```diff
@@ -490,6 +496,63 @@ BEGIN {    # Suppress load of all of these at earliest point.
         return 2;
     }
 
+    sub _blocker_ack_postgresql_datadir ($self) {
+
+        return 0 unless Cpanel::Pkgr::is_installed('postgresql-server');
+
+        my $touch_file = POSTGRESQL_ACK_TOUCH_FILE;
+        return 0 if -e $touch_file;
+
+        my @users_with_dbs = $self->_has_mapped_postgresql_dbs();
+        return 0 unless scalar @users_with_dbs;
+
+        my $message = <<~"EOS";
+    One or more users on your system have associated PostgreSQL databases.
+    ELevate may upgrade the software packages associated with PostgreSQL
+    automatically, but if it does, it will *NOT* automatically update the
+    PostgreSQL data directory to work with the new version. Without an update
+    to the data directory, the upgraded PostgreSQL software will not start, in
+    order to ensure that your data does not become corrupted.
+
+    For more information about PostgreSQL upgrades, please consider the
+    following resources:
+
+    https://cpanel.github.io/elevate/blockers/
+    https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/deploying_different_types_of_servers/using-databases#migrating-to-a-rhel-8-version-of-postgresql_using-postgresql
+    https://www.postgresql.org/docs/10/pgupgrade.html
+
+    When you are ready to acknowledge that you have prepared to update the
+    PostgreSQL data directory, or that this warning does not apply to you,
+    please touch the following file to continue with the ELevate process:
+
+    > touch $touch_file
+
+    The following user(s) have PostgreSQL databases associated with their cPanel accounts:
+    EOS
+
+        $message .= join "\n", @users_with_dbs;
+
+        return $self->has_blocker($message);
+    }
+
+    sub _has_mapped_postgresql_dbs ($self) {
+
+        my $users_json   = Cpanel::SafeRun::Simple::saferunnoerror(qw(/usr/local/cpanel/bin/whmapi1 --output=json list_users));
+        my $users_parsed = Cpanel::JSON::Load($users_json);
+        die "WHM API1 list_users call returned result 0: $users_parsed->{metadata}->{reason}" unless $users_parsed->{metadata}->{result};
+        my @users = grep { $_ ne "root" } $users_parsed->{data}->{users}->@*;
+
+        my @users_with_dbs;
+        foreach my $user (@users) {
+            my $dbs_json   = Cpanel::SafeRun::Simple::saferunnoerror( "/usr/local/cpanel/bin/uapi", "--user=$user", qw(--output=json Postgresql list_databases) );
+            my $dbs_parsed = Cpanel::JSON::Load($dbs_json);
+            die( "UAPI Postgresql::list_databases call returned status 0 for user $user:\n" . join( "\n", $dbs_parsed->{result}->{errors}->@* ) ) unless $dbs_parsed->{result}->{status};
```

Comment:
Should we ignore output like `This server does not support this functionality.`
we are checking for the package earlier, but if it happens, I do not think we should die?

```
{
   "apiversion" : 3,
   "func" : "list_databases",
   "module" : "Postgresql",
   "result" : {
      "data" : null,
      "errors" : [
         "This server does not support this functionality."
      ],
      "messages" : null,
      "metadata" : {},
      "status" : 0,
      "warnings" : null
   }
}
```

Reply:
Just mind localization if we check the error string

Reply:
@atoomic: The translation issue Dan mentioned does complicate things. I don't see a way to force the `uapi` binary to use a particular locale to override the cPanel user's choice. I could call into cPanel internals directly if you think this is important enough.

Reply:
Specifically, something like this:

```perl
sub _has_mapped_postgresql_dbs ($self) {

    # Get a list of cPanel users:
    my %trueuserowners;
    Cpanel::Config::LoadUserOwners::loadtrueuserowners( \%trueuserowners, 1, 1 );
    my @users = keys %trueuserowners;

    # Compile a list of users with PgSQL DBs:
    my @users_with_dbs;
    foreach my $user (@users) {
        my $map = Cpanel::DB::Map::Reader->new( cpuser => $user, engine => 'postgresql' );
        push @users_with_dbs, $user if int $map->get_databases() > 0;
    }

    return @users_with_dbs;
}
```

Reply:
Cannot we set env variables to enforce english?
```
local @ENV{qw{LANG LANGUAGE LC_ALL LC_MESSAGES LC_CTYPE}} = qw{C C C C C};
```

Reply:
I tried several of those at the command line. I don't think the command line API interfaces respect these:

```json
[root@10-2-69-42 ~]# ( export LANG=C LANGUAGE=C LC_ALL=C LC_MESSAGES=C LC_CTYPE=C LOCALE=en CPANEL_SERVER_LOCALE=en ; uapi --user=cptest --output=jsonpretty Postgresql list_databases )
{
   "func" : "list_databases",
   "apiversion" : 3,
   "result" : {
      "metadata" : {},
      "status" : 0,
      "messages" : null,
      "warnings" : null,
      "data" : null,
      "errors" : [
         "Este servidor no soporta esta funcionalidad."
      ]
   },
   "module" : "Postgresql"
}
```

Reply:
what about using `HTTP_COOKIE='session_locale=en'` :-) env variable

Reply:
french
```
╰─> HTTP_COOKIE='session_locale=fr' uapi --user=cptest --output=jsonpretty Postgresql list_databases | jq '.result.errors[0]'
"Le système n’a pas pu se connecter au serveur PostgreSQL « localhost » en raison d’une erreur : (XID dhhfrw) The system failed to connect to the “PostgreSQL” database “postgres” because of an error: connection_failure (FATAL:  password authentication failed for user \"postgres\")\nLe système n’a pas pu se connecter au serveur PostgreSQL « localhost » en raison d’une erreur : (XID dhhfrw) The system failed to connect to the “PostgreSQL” database “postgres” because of an error: connection_failure (FATAL:  password authentication failed for user \"postgres\") at /usr/local/cpanel/Cpanel/PostgresAdmin/Basic.pm line 284.\n"
```
spanish...
```
╰─> HTTP_COOKIE='session_locale=es' uapi --user=cptest --output=jsonpretty Postgresql list_databases | jq '.result.errors[0]'
"El sistema no se pudo conectar con el servidor PostgreSQL “localhost” debido a un error: (XID sqeq4g) The system failed to connect to the “PostgreSQL” database “postgres” because of an error: connection_failure (FATAL:  password authentication failed for user \"postgres\")\nEl sistema no se pudo conectar con el servidor PostgreSQL “localhost” debido a un error: (XID sqeq4g) The system failed to connect to the “PostgreSQL” database “postgres” because of an error: connection_failure (FATAL:  password authentication failed for user \"postgres\") at /usr/local/cpanel/Cpanel/PostgresAdmin/Basic.pm line 284.\n"
```

thus we can enforce English :-)

Reply:
The function now short-circuits if it ever sees `This server does not support this functionality.` It still dies on other failures.